types/src/transaction.rs

```diff
@@ -227,13 +230,104 @@ impl IntoProto for RawTransaction {
     }
 }
 
+impl CanonicalSerialize for RawTransaction {
+    fn serialize(&self, serializer: &mut impl CanonicalSerializer) -> Result<()> {
+        serializer.encode_struct(&self.sender)?;
+        serializer.encode_u64(self.sequence_number)?;
+        serializer.encode_struct(&self.payload)?;
+        serializer.encode_u64(self.max_gas_amount)?;
+        serializer.encode_u64(self.gas_unit_price)?;
+        serializer.encode_u64(self.expiration_time.as_secs())?;
+        Ok(())
+    }
+}
+
+impl CanonicalDeserialize for RawTransaction {
+    fn deserialize(deserializer: &mut impl CanonicalDeserializer) -> Result<Self> {
+        let sender = deserializer.decode_struct()?;
+        let sequence_number = deserializer.decode_u64()?;
+        let payload = deserializer.decode_struct()?;
+        let max_gas_amount = deserializer.decode_u64()?;
+        let gas_unit_price = deserializer.decode_u64()?;
+        let expiration_time = Duration::from_secs(deserializer.decode_u64()?);
+
+        Ok(RawTransaction {
+            sender,
+            sequence_number,
+            payload,
+            max_gas_amount,
+            gas_unit_price,
+            expiration_time,
+        })
+    }
+}
+
 #[derive(Clone, Debug, Hash, Eq, PartialEq, Serialize, Deserialize)]
 pub enum TransactionPayload {
     /// A regular programmatic transaction that is executed by the VM.
     Program(Program),
     WriteSet(WriteSet),
 }
 
+impl CanonicalSerialize for TransactionPayload {
+    fn serialize(&self, serializer: &mut impl CanonicalSerializer) -> Result<()> {
+        match self {
+            TransactionPayload::Program(program) => {
+                serializer.encode_u32(TransactionPayloadType::Program as u32)?;
+                serializer.encode_struct(program)?;
+            }
+            TransactionPayload::WriteSet(write_set) => {
+                serializer.encode_u32(TransactionPayloadType::WriteSet as u32)?;
+                serializer.encode_struct(write_set)?;
+            }
+        };
+        Ok(())
+    }
+}
+
+impl CanonicalDeserialize for TransactionPayload {
+    fn deserialize(deserializer: &mut impl CanonicalDeserializer) -> Result<Self> {
+        let decoded_payload_type = deserializer.decode_u32()?;
+        let payload_type = TransactionPayloadType::from_u32(decoded_payload_type);
+        let payload_type: TransactionPayloadType = match payload_type {
```

Comment:
This can be rewritten as an `unwrap_or_else`, or an [`if let`](https://doc.rust-lang.org/rust-by-example/flow_control/if_let.html)

Reply:
This is fine too:
```
       let payload_type: TransactionPayloadType = match payload_type {
 @huitseeker
huitseeker 3 hours ago Member
This can be rewritten as an unwrap_or_else, or an if let

@huitseeker	Reply…
            Some(v) => v,
            None => {
                return Err(format_err!(
                    "ParseError: Unable to decode TransactionPayloadType, found {}",
                    decoded_payload_type
                ))
            }
        };

        Ok(match payload_type {
            TransactionPayloadType::Program => {
                TransactionPayload::Program(deserializer.decode_struct()?)
            }
            TransactionPayloadType::WriteSet => {
                TransactionPayload::WriteSet(deserializer.decode_struct()?)
            }
        })
    }
}
```

Reply:
I believe you can also replace the whole end section of this with
```
Err(format_err!(
                    "ParseError: Unable to decode TransactionPayloadType, found {}",
                    decoded_payload_type
                ))
```

(untested, FYI, YMMV, etc)